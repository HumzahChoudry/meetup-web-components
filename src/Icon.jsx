import React from 'react';
import cx from 'classnames';

export const ICON_CLASS = 'svg';

export const MEDIA_SIZES = {
	xs: '16',
	s: '24',
	m: '36',
	l: '48',
	xl: '72',
	auto: '100%',
};

/**
 * Icon component used to insert an svg icon into a component or page
 *
 * **Accessibility** If an Icon is used on its own without supporting
 * text to explain what it is/does, be a good citizen and pass in an
 * `aria-label` attribute describing what the icon represents
 *
 * @module Icon
 */
class Icon extends React.PureComponent {

	render() {
		const {
			className,
			shape,
			size,
			...other
		} = this.props;

		const classNames = cx(
			ICON_CLASS,
			`svg--${shape}`,
			className
		);

		const viewBox = size === 'auto' ? MEDIA_SIZES['xl'] : MEDIA_SIZES[size];
		const dim = MEDIA_SIZES[size];

		return (
			<span className={classNames}>
				<svg
					preserveAspectRatio='xMinYMin meet'
					width={dim}
					height={dim}
					viewBox={`0 0 ${viewBox} ${viewBox}`}
					className='svg-icon valign--middle'
					role='img'
					{...other}>
					<use xlinkHref={`#icon-${shape}`}></use>
				</svg>
			</span>
		);
	}
}

Icon.defaultProps = {
	size: 'm'
};

Icon.propTypes = {
	shape: React.PropTypes.string.isRequired,
	size: React.PropTypes.oneOf(['xs', 's', 'm', 'l', 'xl', 'auto'])
};

export default Icon;
