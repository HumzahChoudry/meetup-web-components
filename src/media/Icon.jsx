import React from 'react';
import cx from 'classnames';
import { MEDIA_SIZES, MEDIA_QUERIES } from '../utils/designConstants';

export const ICON_CLASS = 'svg';

/**
 * Icon component used to insert an svg icon into a component or page
 *
 * **Accessibility** If an Icon is used on its own without supporting
 * text to explain what it is/does, be a good citizen and pass in an
 * `aria-label` attribute describing what the icon represents
 *
 * @module Icon
 */
class Icon extends React.PureComponent {
	constructor(props){
		super(props);

		this.state = {
			iconScaleFactor: undefined
		};
	}

	componentDidMount() {

		this.setupListeners = (name, mediaQuery) => {
			if (!window.matchMedia) return;

			const mql = window.matchMedia(mediaQuery);
			mql._fn = function (e) {
				return this.handleMediaChange(e.matches, name);
			}.bind(this);
			mql.addListener(mql._fn);

			return mql;
		};

		this.handleMediaChange = (matches, name) => {
			if (matches) {
				this.changeHandler(name);
			}
		};

		this.changeHandler = (name) => {
			let scaleFactor;

			switch (name) {

			case 'small':
				scaleFactor = 1;
				// console.log('do a small thing');
				break;

			case 'medium':
				scaleFactor = 1.125;
				// console.log('do a medium thing');
				break;

			case 'large':
				scaleFactor = 1.25;
				// console.log('do a large thing');
				break;

			case 'huge':
				scaleFactor = 1.25;
				// console.log('do a huge thing');
				break;

			default:
				break;
			}

			this.setState({
				iconScaleFactor: scaleFactor
			});
		};

		// where the stuff gets called
		for(const mq in MEDIA_QUERIES) {
			this.setupListeners(mq, MEDIA_QUERIES[mq]);

			if (window.matchMedia(MEDIA_QUERIES[mq]).matches) {
				this.changeHandler(mq);
			}
		}

	}

	render() {
		const {
			className,
			shape,
			size,
			...other
		} = this.props;

		const classNames = cx(
			ICON_CLASS,
			`svg--${shape}`,
			className
		);

		const viewBox = size === 'auto' ? MEDIA_SIZES['xl'] : MEDIA_SIZES[size];
		const dim = MEDIA_SIZES[size] * this.state.iconScaleFactor;

		return (
			<span className={classNames}>
				<svg
					preserveAspectRatio='xMinYMin meet'
					width={dim}
					height={dim}
					viewBox={`0 0 ${viewBox} ${viewBox}`}
					className='svg-icon valign--middle'
					role='img'
					{...other}>
					<use xlinkHref={`#icon-${shape}`}></use>
				</svg>
			</span>
		);
	}
}

Icon.defaultProps = {
	size: 'm'
};

Icon.propTypes = {
	shape: React.PropTypes.string.isRequired,
	size: React.PropTypes.oneOf(['xs', 's', 'm', 'l', 'xl', 'auto'])
};

export default Icon;
