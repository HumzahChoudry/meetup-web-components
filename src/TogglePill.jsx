import React from 'react';
import cx from 'classnames';
import Icon from './Icon';

export const TOGGLE_PILL_CLASS = 'toggleButton';

/**
 * SQ2 Toggle Pill component
 * @see {@link https://github.com/meetup/sassquatch2/blob/develop/sass/ui-components/_toggle-pill.scss}
 * @see {@link http://meetup.github.io/sassquatch2/ui_components.html#togglePills}
 * @module TogglePill
 */
export default class TogglePill extends React.Component {
	constructor (props) {
		super(props);

		this.state = {isChecked: props.checked || false};

		this.onChange = this.onChange.bind(this);
	}

	onChange(e) {
		this.setState({isChecked: !this.state.isChecked});

		if (this.props.onChange) {
			this.props.onChange(e);
		}
	}

	render() {
		const {
			id,
			name,
			value,
			children,
			className,
			topic,
			...other
		} = this.props;

		delete other.onChange; // onChange is consumed in this.onChange - do not pass it along to children

		const classNames = cx(
			TOGGLE_PILL_CLASS,
			{
				'toggleButton--topic': topic
			},
			className
		);

		// ---
		// Topic variant
		const topicClassName = cx(
			'toggleButton-icon',
			{
				'toggleButton-icon--active' : this.state.isChecked,
				'toggleButton-icon--inactive' : (!this.state.isChecked)
			}
		);

		const topicShape = (this.state.isChecked) ? 'heart' : 'heart-outline';

		const topicChildren = (
			<Icon
				className={topicClassName}
				shape={topicShape}
				size='xs'
				label='Active Topic Pill Icon'/>
		);
		// ---

		return (
			<div className={classNames}>
				<input
					className='toggleButton-input'
					type='checkbox'
					id={id}
					name={name}
					value={value}
					checked={this.state.isChecked}
					onChange={this.onChange}
					{...other} />
				<label
					className='toggleButton-label'
					htmlFor={id}>
						{children}
						{(topic) ? topicChildren : null}
				</label>
			</div>
		);
	}
}

TogglePill.protoTypes = {
	id: React.PropTypes.string.isRequired,
	name: React.PropTypes.string.isRequired,
	value: React.PropTypes.string.isRequired,
	children: React.PropTypes.node.isRequired,
	checked: React.PropTypes.bool,
	topic: React.PropTypes.bool
};
TogglePill.defaultProps = {
	checked: false
};

