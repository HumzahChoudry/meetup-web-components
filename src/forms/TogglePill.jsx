import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';
import Icon from '../media/Icon';
import withToggleControl from '../utils/WithToggleControl';

export const TOGGLE_BUTTON_CLASS = 'toggleButton';

/**
 * Toggle Pill component
 * @see {@link https://github.com/meetup/sassquatch2/blob/develop/sass/ui-components/_toggle-pill.scss}
 * @see {@link http://meetup.github.io/sassquatch2/ui_components.html#togglePills}
 * @module TogglePill
 */
class TogglePill extends React.Component {
	constructor (props) {
		super(props);

		this.onChange = this.onChange.bind(this);
	}

	onChange(e) {
		this.props.toggleActive();

		if (this.props.onChange) {
			this.props.onChange(e);
		}
	}

	render() {
		const {
			isActive,
			children,
			className,
			topic,
			id,
			name,
			useRadio,
			value,
			toggleActive, // eslint-disable-line no-unused-vars
			...other
		} = this.props;

		delete other.onChange; // onChange is consumed in this.onChange - do not pass it along to children

		const inputType = useRadio ? 'radio' : 'checkbox';

		const classNames = cx(
			TOGGLE_BUTTON_CLASS,
			{
				'toggleButton--topic': topic,
				'toggleButton--radio': useRadio
			},
			className
		);

		const labelClassNames = cx(
			'toggleButton-label',
			{
				'toggleButton-label--radio': useRadio
			}
		);

		// ---
		// Topic variant
		const topicClassName = cx(
			'toggleButton-icon',
			{
				'toggleButton-icon--active' : isActive,
				'toggleButton-icon--inactive' : (!isActive)
			}
		);

		const iconShape = isActive ? 'heart' : 'heart-outline';

		const topicChildren = (
			<Icon
				className={topicClassName}
				shape={iconShape}
				size='xxs'
				label='Active Topic Icon'/>
		);
		// ---

		return (
			<div className={classNames}>
				<input
					className='toggleButton-input visibility--a11yHide'
					type={inputType}
					id={id}
					name={name}
					value={value}
					checked={isActive}
					onChange={this.onChange}
					{...other} />
				<label
					className={labelClassNames}
					htmlFor={id}>
					{children}
					{(topic) ? topicChildren : null}
				</label>
			</div>
		);
	}
}

TogglePill.protoTypes = {
	id: PropTypes.string.isRequired,
	name: PropTypes.string.isRequired,
	value: PropTypes.string.isRequired,
	children: PropTypes.node.isRequired,
	isActive: PropTypes.bool,
	topic: PropTypes.bool
};
TogglePill.defaultProps = {
	isActive: false
};

export default withToggleControl(TogglePill);

