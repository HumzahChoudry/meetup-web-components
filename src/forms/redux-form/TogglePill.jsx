import React from 'react';
import TogglePill from '../TogglePill';

/**
 * @param {Object} props - props passed in from parent and redux-form
 * @description wraps standard TogglePill web component for use with redux-form
 * deconstructs props that redux-forms sets and sets them on TextInput
 * @return {Component} TogglePill
 */
const ReduxFormTogglePill = props => {
	const {
		meta, // eslint-disable-line no-unused-vars
		// not adding meta error to TogglePill prop for now
		input,
		...other
	} = props;

	console.log('HELLO PROPS');
	// const checkedProp = input.value == true ? 'checked' : '';

	return <TogglePill {...input} isActive={!!input.value} {...other} />;
};

ReduxFormTogglePill.displayName = 'ReduxFormTogglePill';

export default ReduxFormTogglePill;
