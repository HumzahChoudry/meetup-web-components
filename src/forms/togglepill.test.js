import React from 'react';
import ReactDOM from 'react-dom';
import TestUtils from 'react-addons-test-utils';
import TogglePill, { TOGGLE_PILL_CLASS } from './TogglePill';
import Icon from '../media/Icon';

describe('TogglePill', () => {
	const id = 'togglePillId',
		name = 'togglePillName',
		label = 'Toggle Pill Label',
		value = 'Toggle Pill Value';

	let togglePillEl;

	beforeEach(() => {
		const togglePill = TestUtils.renderIntoDocument(
			<TogglePill
				id={id}
				name={name}
				value={value}>
				{label}
			</TogglePill>
		);

		togglePillEl = ReactDOM.findDOMNode(togglePill);
	});

	afterEach(() => {
		togglePillEl = null;
	});

	it('exists', () => {
		expect(togglePillEl).not.toBeNull();
	});

	it('has SQ2 toggle-pill styles', () => {
		expect(togglePillEl.classList.contains(TOGGLE_PILL_CLASS)).toBe(true);
	});

	it('creates an input element', () => {
		const len = togglePillEl.getElementsByTagName('INPUT').length;
		expect(len).toBe(1);
	});

	it('creates a label element', () => {
		const len = togglePillEl.getElementsByTagName('LABEL').length;
		expect(len).toBe(1);
	});

	it('executes onChange when clicked', () => {
		const spyable = {
			onChange: (e) => {}
		};

		spyOn(spyable, 'onChange');
		const togglePill = TestUtils.renderIntoDocument(
			<TogglePill
				onChange={spyable.onChange}
				id={id}
				name={name}
				value={value}>
				{label}
			</TogglePill>);
		const togglePillEl = ReactDOM.findDOMNode(togglePill);

		expect(togglePillEl.getElementsByTagName('input').length).toEqual(1);
		const input = togglePillEl.getElementsByTagName('input')[0];

		TestUtils.Simulate.change(input);

		expect(spyable.onChange).toHaveBeenCalled();
	});


	describe('TogglePill with icon', () => {
		const id = 'togglePillId',
			name = 'togglePillName',
			label = 'Toggle Pill Label',
			value = 'Toggle Pill Value';

		let togglePill;

		beforeEach(() => {
			togglePill = TestUtils.renderIntoDocument(
				<TogglePill
					topic
					id={id}
					name={name}
					value={value}>
					{label}
				</TogglePill>
			);
		});


		afterEach(() => {
			togglePill = null;
		});


		it('creates an Icon component', function() {
			const len = TestUtils.scryRenderedComponentsWithType(togglePill, Icon).length;
			expect(len).toBeGreaterThan(0);
		});

	});
});
