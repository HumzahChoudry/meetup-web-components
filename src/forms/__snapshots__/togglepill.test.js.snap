// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TogglePill renders a component with expected attributes 1`] = `
<WithToggle
  id="hikingCategory"
  isActive={false}
  name="meetupCategories"
  value="hiking"
>
  <span
    aria-pressed={false}
    role="button"
  >
    <TogglePill
      id="hikingCategory"
      isActive={false}
      name="meetupCategories"
      toggleActive={[Function]}
      value="hiking"
    >
      <div
        className="toggleButton"
      >
        <input
          checked={false}
          className="toggleButton-input visibility--a11yHide"
          id="hikingCategory"
          name="meetupCategories"
          onChange={[Function]}
          type="checkbox"
          value="hiking"
        />
        <label
          className="toggleButton-label"
          htmlFor="hikingCategory"
        >
          Hiking!
        </label>
      </div>
    </TogglePill>
  </span>
</WithToggle>
`;
